### PR TITLE
feat: add dataset_format parameter to dataset_download component

### PR DESCRIPTION
**Pull request by ChughShilpa**
## Summary
- Adds an optional `dataset_format` parameter (default `"chat"`) to the `dataset_download` component to support GRPO tool-call dataset validation
- When `dataset_format="tool_call"`, validates single-turn (`target_tool_name` + `question`) or multi-turn (`messages` with assistant `tool_calls`) tool-call datasets
- Existing SFT/LoRA/OSFT pipelines are unaffected — they use the default `"chat"` format with zero changes

## Test plan
- [x] 17 unit tests pass (5 existing + 12 new covering both formats, edge cases, mixed format rejection)
- [x] All 12 finetuning pipeline compile tests pass (SFT, LoRA, OSFT × standard/minimal)
- [x] `ruff check` and `ruff format --check` pass
- [x] Backward compatible — no changes to existing pipeline files

Ref: [RHOAIENG-90696](https://redhat.atlassian.net/browse/RHOAIENG-90696)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

[RHOAIENG-90696]: https://redhat.atlassian.net/browse/RHOAIENG-90696?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary by CodeRabbit

* **New Features**
  * Added support for selecting between chat-template and tool-call dataset formats, with chat as the default.
  * Added validation for GRPO tool-call datasets in single-turn and multi-turn formats.
  * Tool-call validation checks assistant tool calls and supports datasets containing up to 100 samples.

* **Bug Fixes**
  * Improved handling of empty, malformed, unsupported, or mixed-format datasets.
  * Preserved existing chat-format validation behavior.

* **Documentation**
  * Documented the supported dataset formats and the new format selection option.

### CODE REVIEW

**Comment by efazal** [comment]
Thanks @ChughShilpa — the direction looks good overall. I left two comments on malformed tool-call acceptance and unsupported-format test coverage.

File: components/data_processing/dataset_download/component.py [hunk 1]
```diff
@@ -131,6 +139,75 @@ def validate_chat_format_dataset(dataset: Dataset) -> bool:
         log_message(f"Dataset validated: {len(dataset)} examples in chat format")
         return True
 
+    def _has_tool_calls_in_messages(messages: list) -> bool:
+        """Check if any assistant message in the list contains tool_calls."""
+        for msg in messages:
+            if isinstance(msg, dict) and msg.get("role") == "assistant" and "tool_calls" in msg:
```
Comment (efazal) [inline components/data_processing/dataset_download/component.py]: Checking only for the key lets tool_calls=[]/None/scalars pass as valid tool-call data, so malformed traces can be written for GRPO. Require a non-empty list and add regression cases for these values.

File: components/data_processing/dataset_download/tests/test_component_unit.py [hunk 2]
```diff
@@ -94,3 +168,219 @@ def test_component_supports_multiple_uri_schemes(self):
...
+            [
+                {"messages": [{"role": "user", "content": "Hello"}, {"role": "assistant", "content": "Hi"}]},
+                {"messages": [{"role": "user", "content": "Bye"}, {"role": "assistant", "content": "Goodbye"}]},
+            ]
+        )
+        assert self.validate_chat(dataset) is True
+
+    def test_invalid_dataset_format_value(self):
+        """Passing unsupported dataset_format value raises ValueError."""
+        sig = inspect.signature(dataset_download.python_func)
```
Comment (efazal) [inline components/data_processing/dataset_download/tests/test_component_unit.py]: nit: this test never passes an unsupported dataset_format; it only reasserts the default, so the ValueError dispatch branch is untested. Exercise that branch directly so this regression test can fail when validation is removed.

File: components/data_processing/dataset_download/component.py [hunk 3]
```diff
@@ -131,6 +139,75 @@ def validate_chat_format_dataset(dataset: Dataset) -> bool:
...
+                        f"Item {i} missing 'target_tool_name' field. "
+                        f"All samples must be single-turn format (detected from first sample). "
+                        f"Found keys: {list(item.keys())}"
+                    )
+                if "question" not in item:
+                    raise ValueError(
+                        f"Item {i} missing 'question' field. "
+                        f"All samples must be single-turn format (detected from first sample). "
+                        f"Found keys: {list(item.keys())}"
+                    )
```
Comment (Sridhar1030) [inline components/data_processing/dataset_download/component.py]: These two checks never fire on a real `datasets.Dataset`. It is Arrow backed with one shared schema, so every row carries every column and a missing value comes back as `None`, not as an absent key:

```
input:  [{"question":"CA?","target_tool_name":"get_alerts"}, {"question":"NY?"}]
ds[1]                        -> {'question': 'NY?', 'target_tool_name': None}
"target_tool_name" in ds[1]  -> True
```

Row 1 has no tool name in the source file, but validation passes and `target_tool_name: None` reaches training. Checking the value also covers empty strings.

```suggestion
                if not item.get("target_tool_name"):
                    raise ValueError(
                        f"Item {i}: 'target_tool_name' is missing or empty. "
                        f"All samples must be single-turn format (detected from first sample)."
                    )
                if not item.get("question"):
                    raise ValueError(
                        f"Item {i}: 'question' is missing or empty. "
                        f"All samples must be single-turn format (detected from first sample)."
                    )
```

`_MockDataset` is a plain list of dicts so it does not show this. Building those fixtures with `Dataset.from_list(...)` would turn `test_single_turn_missing_target_tool_name`, `test_single_turn_missing_question` and `test_mixed_formats_rejected` red against the current code.

File: components/data_processing/dataset_download/component.py [hunk 4]
```diff
@@ -338,9 +415,15 @@ def load_from_local(file_path: str) -> Dataset:
...
-        # Validate chat template format
-        log_message("Validating chat template format...")
-        validate_chat_format_dataset(dataset)
+        # Validate dataset format
+        if dataset_format == "chat":
+            log_message("Validating chat template format...")
+            validate_chat_format_dataset(dataset)
+        elif dataset_format == "tool_call":
+            log_message("Validating tool-call format...")
+            validate_tool_call_format_dataset(dataset)
```
Comment (Sridhar1030) [inline components/data_processing/dataset_download/component.py:440]: This path writes corrupt output. Ran `dataset_download.python_func` with a 10 row multi-turn file and `dataset_format="tool_call"`. Validation passes, the run reports success, and `train.jsonl` line 0 is:

```
messages[0] type: str
messages[0] value: '{"role":"user","content":"q6"}'
```

Same run with `dataset_format="chat"` gives `dict`, so only the new path is affected.

Multi-turn messages are not uniform (user has role/content, assistant adds tool_calls, tool adds tool_call_id), so Arrow cannot infer a struct and types the column `List(Json(decode=True))`. The write at [component.py#L445](https://github.com/opendatahub-io/pipelines-components/blob/89eb3175e5d14553a3ed710e4a115d2904fb2df2/components/data_processing/dataset_download/component.py#L445-L448) then re-emits each message as an encoded string. Anything downstream doing `messages[i]["role"]` hits a TypeError.

Reproduced on `datasets` 4.8.4 locally, not inside the base image. The typing is version dependent, so worth confirming against the version pinned in `odh-th06-cpu-torch291-py312:odh-3.4`.

A test that calls `python_func` and reads the output file back would catch this, and would also cover this dispatch, which currently has no test on either branch.

Comment (ChughShilpa) [reply components/data_processing/dataset_download/component.py:440]: Seems a real concern. Thanks for pointing it out
I added a validation branch for new dataset format and issue is that the existing save path wasn't designed for non-uniform message schemas.

**Comment by efazal** [comment]
Thanks @ChughShilpa, the validation logic and test coverage for both formats look good. A few things to address before merge.

File: components/data_processing/dataset_download/component.py [hunk 5]
```diff
@@ -338,9 +417,15 @@ def load_from_local(file_path: str) -> Dataset:
...
```
Comment (efazal) [inline components/data_processing/dataset_download/component.py:440]: Validation passes here but the written output may be corrupt: when tool-call messages have heterogeneous keys (user: `{role, content}`, assistant: `{role, content, tool_calls}`), datasets ≥4.x infers the column as `List(string)` — HF auto-decodes on row access so validation sees dicts, but `to_json()` writes each message as an encoded string, so downstream `messages[i]["role"]` hits a `TypeError`.
Fix: after loading and before this call, normalize with `dataset = dataset.map(lambda r: {**r, "messages": [json.loads(m) if isinstance(m, str) else m for m in (r.get("messages") or [])]})`, and add a test that reads the written JSONL back and asserts `isinstance(messages[0], dict)`.

File: components/data_processing/dataset_download/component.py [hunk 6]
```diff
@@ -338,9 +417,15 @@ def load_from_local(file_path: str) -> Dataset:
             else:
                 log_message(f"Subset count ({subset_count}) >= dataset size ({original_size}), using all examples")
 
-        # Validate chat template format
-        log_message("Validating chat template format...")
-        validate_chat_format_dataset(dataset)
+        # Validate dataset format
+        if dataset_format == "chat":
```
Comment (efazal) [inline components/data_processing/dataset_download/component.py:435]: `dataset_format` is only checked after the full dataset is downloaded — a typo like `"toolcall"` triggers a potentially multi-minute download before failing.
Add a guard at the top of the component body, before `parse_uri`, to reject unknown format values immediately.

File: components/data_processing/dataset_download/component.py [hunk 7]
```diff
@@ -131,6 +139,77 @@ def validate_chat_format_dataset(dataset: Dataset) -> bool:
...
+
+        if not is_single_turn and not is_multi_turn:
+            raise ValueError(
+                f"Item 0 does not match any supported tool-call format. "
+                f"Expected either (1) single-turn with 'target_tool_name' and 'question' fields, "
+                f"or (2) multi-turn with 'messages' containing assistant tool_calls. "
+                f"Found keys: {list(first.keys())}"
+            )
+
+        if is_single_turn:
```
Comment (efazal) [inline components/data_processing/dataset_download/component.py:190]: nit: when an Arrow-backed dataset has all four relevant column names (`target_tool_name`, `question`, `messages`, plus an assistant `tool_calls` entry), both `is_single_turn` and `is_multi_turn` evaluate to `True` and single-turn silently wins.
A `log_message` on the tie-break path, or an explicit error if both are true, would make this visible in logs rather than a silent choice.

File: components/data_processing/dataset_download/component.py (hunk 4 above)
Comment (ChughShilpa) [reply components/data_processing/dataset_download/component.py:440]: Investigated — doesn't reproduce on datasets 4.3.0 (base image) or 4.8.4. 
Tested on datasets 4.8.4 with 10 multi-turn rows including user, assistant+tool_calls, tool+tool_call_id, and final assistant messages. Arrow infers a unified struct and to_json() writes all messages as dicts — not strings.
@Sridhar1030 
Happy to dig deeper if we can reproduce it.

**Comment by efazal** [comment]
The corrupt output issue is still present. The `to_json()` calls at lines 453/456/466/469 are unchanged. When a multi-turn tool-call dataset is loaded through the `datasets` library, Arrow infers a union schema and pads every message struct with all fields — so user messages end up with `"tool_calls": null` written to disk. Sridhar confirmed this empirically.

The validation added in this PR correctly rejects bad input, but a valid multi-turn dataset still produces malformed output at the save step.

Fix: replace all four `to_json()` calls with a helper that strips `None` values from message dicts before writing:

```python
        def _write_jsonl(ds, path):
            import json
            with open(path, "w") as f:
                for row in ds:
                    if dataset_format == "tool_call" and "messages" in row:
                        row = dict(row)
                        row["messages"] = [
                            {k: v for k, v in m.items() if v is not None}
                            for m in (row["messages"] or [])
                        ]
                    f.write(json.dumps(row) + "\n")

        log_message(f"Saving train dataset to {train_dataset.path}")
        _write_jsonl(train_ds, train_dataset.path)

        log_message(f"Saving eval dataset to {eval_dataset.path}")
        _write_jsonl(eval_ds, eval_dataset.path)

        pvc_dataset_dir = os.path.join(pvc_mount_path, "datasets")
        os.makedirs(pvc_dataset_dir, exist_ok=True)
        pvc_train_path = os.path.join(pvc_dataset_dir, "train.jsonl")
        pvc_eval_path = os.path.join(pvc_dataset_dir, "eval.jsonl")

        log_message(f"Saving train dataset to PVC: {pvc_train_path}")
        _write_jsonl(train_ds, pvc_train_path)

        log_message(f"Saving eval dataset to PVC: {pvc_eval_path}")
        _write_jsonl(eval_ds, pvc_eval_path)
```

A regression test worth adding: load a two-row multi-turn JSONL where user messages have no `tool_calls` key, run the component end-to-end (or call `python_func` directly), read back the written JSONL, and assert that no user message row contains a `tool_calls` key.

File: components/data_processing/dataset_download/component.py (hunk 4 above)
Comment (Sridhar1030) [reply components/data_processing/dataset_download/component.py:440]: Thanks for digging in. Reproduced without the component:

```python
import json, datasets
row = {"messages": [
    {"role": "user", "content": "hi"},
    {"role": "assistant", "content": None, "tool_calls": [{"id": "c1", "type": "function", "function": {"name": "f", "arguments": "{}"}}]},
    {"role": "tool", "content": "ok", "tool_call_id": "c1"},
]}
open("in.jsonl", "w").write("\n".join([json.dumps(row)] * 10))
datasets.load_dataset("json", data_files="in.jsonl", split="train").to_json("out.jsonl", orient="records", lines=True)
print(datasets.__version__, type(json.loads(open("out.jsonl").readline())["messages"][0]).__name__)
```

`datasets` 4.8.4 prints `str`, 4.3.0 prints `dict`. On 4.8.4, `Dataset.from_list` writes dicts, which may explain your run.

Not blocking, since the base image is fine today. A test that reads the output back would catch this on a `datasets` upgrade.

Comment (ChughShilpa) [reply components/data_processing/dataset_download/component.py:440]: Thanks for the isolated repro — that matches the `load_dataset("json")` path, and `Dataset.from_list` was why my earlier run didn't show it.

Addressed in `_write_jsonl`: if a message comes back as a string we `json.loads` it, and we strip `None` fields Arrow added during schema unification. The round-trip test now calls `dataset_download.python_func` on a two-row multi-turn JSONL and reads the written file back, so this fails if the helper is removed. Not an issue on the current base image (`datasets` 4.3.0), but this keeps the upgrade path safe.

**Comment by ChughShilpa** [comment]
> The corrupt output issue is still present. The `to_json()` calls at lines 453/456/466/469 are unchanged. When a multi-turn tool-call dataset is loaded through the `datasets` library, Arrow infers a union schema and pads every message struct with all fields — so user messages end up with `"tool_calls": null` written to disk. Sridhar confirmed this empirically.
> 
> The validation added in this PR correctly rejects bad input, but a valid multi-turn dataset still produces malformed output at the save step.
> 
> Fix: replace all four `to_json()` calls with a helper that strips `None` values from message dicts before writing:
> 
> ```python
>         def _write_jsonl(ds, path):
>             import json
>             with open(path, "w") as f:
>                 for row in ds:
>                     if dataset_format == "tool_call" and "messages" in row:
>                         row = dict(row)
>                         row["messages"] = [
>                             {k: v for k, v in m.items() if v is not None}
>                             for m in (row["messages"] or [])
>                         ]
>                     f.write(json.dumps(row) + "\n")
> 
>         log_message(f"Saving train dataset to {train_dataset.path}")
>         _write_jsonl(train_ds, train_dataset.path)
> 
>         log_message(f"Saving eval dataset to {eval_dataset.path}")
>         _write_jsonl(eval_ds, eval_dataset.path)
> 
>         pvc_dataset_dir = os.path.join(pvc_mount_path, "datasets")
>         os.makedirs(pvc_dataset_dir, exist_ok=True)
>         pvc_train_path = os.path.join(pvc_dataset_dir, "train.jsonl")
>         pvc_eval_path = os.path.join(pvc_dataset_dir, "eval.jsonl")
> 
>         log_message(f"Saving train dataset to PVC: {pvc_train_path}")
>         _write_jsonl(train_ds, pvc_train_path)
> 
>         log_message(f"Saving eval dataset to PVC: {pvc_eval_path}")
>         _write_jsonl(eval_ds, pvc_eval_path)
> ```
> 
> A regression test worth adding: load a two-row multi-turn JSONL where user messages have no `tool_calls` key, run the component end-to-end (or call `python_func` directly), read back the written JSONL, and assert that no user message row contains a `tool_calls` key.

Fixed — all four `to_json()` calls are replaced by `_write_jsonl`. It writes messages as dicts (decoding JSON strings when Arrow typed the column that way) and strips `None` values, so user messages no longer pick up `"tool_calls": null`.

`test_multiturn_messages_written_as_dicts` now runs `python_func` end-to-end on a two-row multi-turn JSONL and asserts messages are dicts and that user messages have no `tool_calls` key.

File: components/data_processing/dataset_download/component.py (hunk 7 above)
Comment (ChughShilpa) [reply components/data_processing/dataset_download/component.py:190]: Fixed — if item 0 matches both single-turn and multi-turn, we now log that single-turn validation is used instead of choosing silently. Covered by `test_both_formats_logs_tie_break`.

File: components/data_processing/dataset_download/component.py (hunk 1 above)
Comment (ChughShilpa) [reply components/data_processing/dataset_download/component.py]: Fixed — `_has_tool_calls_in_messages` now requires a non-empty list, so `[]` / `None` / scalars no longer count as tool-call data. Regression tests cover all three cases (`test_has_tool_calls_rejects_empty_list`, `test_has_tool_calls_rejects_none`, `test_has_tool_calls_rejects_scalar`).

**Review by efazal — APPROVED** [review]
Thanks @ChughShilpa 
/lgtm
